File: Readio/src/main/java/ninegle/Readio/mypage/service/MyPageUserService.java
package ninegle.Readio.mypage.service;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import lombok.RequiredArgsConstructor;
import ninegle.Readio.user.domain.User;
import ninegle.Readio.user.repository.UserRepository;
import ninegle.Readio.user.service.UserContextService;
import ninegle.Readio.global.exception.BusinessException;
import ninegle.Readio.global.exception.domain.ErrorCode;
import ninegle.Readio.mypage.dto.request.UserUpdateRequestDto;
import ninegle.Readio.mypage.dto.response.UserInfoDto;
import ninegle.Readio.mypage.mapper.MyPageUserMapper;

@Service
@RequiredArgsConstructor
public class MyPageUserService {

	private final UserRepository userRepository;
	private final UserContextService userContextService;

	// 읽기 전용 트랜잭션
	@Transactional(readOnly = true)
	public UserInfoDto getUserInfo() {
		Long userId = userContextService.getCurrentUserId();

		User user = userRepository.findById(userId)
			.orElseThrow(() -> new BusinessException(ErrorCode.USER_NOT_FOUND));

		return MyPageUserMapper.toUserInfoDto(user);
	}

	// 데이터 수정하는 메서드
	@Transactional
	public UserInfoDto updateUserInfo(UserUpdateRequestDto dto) {
		Long userId = userContextService.getCurrentUserId();

		// 닉네임, 전화번호 둘 다 비어있는 경우
		if ((dto.getNickname() == null || dto.getNickname().isBlank()) &&
			(dto.getPhoneNumber() == null || dto.getPhoneNumber().isBlank())) {
			throw new BusinessException(ErrorCode.MISSING_REQUIRED_FIELD);
		}

		User user = userRepository.findById(userId)
			.orElseThrow(() -> new BusinessException(ErrorCode.USER_NOT_FOUND));

		// 닉네임 업데이트 로직
		if (dto.getNickname() != null) {
			validateNickname(dto.getNickname());
			user.updateNickname(dto.getNickname());
		}

		// 전화번호 업데이트 로직
		if (dto.getPhoneNumber() != null) {
			validatePhoneNumber(dto.getPhoneNumber());
			user.updatePhoneNumber(dto.getPhoneNumber());
		}

		return MyPageUserMapper.toUserInfoDto(userRepository.save(user));
	}

	// 닉네임 검증 메서드
	private void validateNickname(String nickname) {
		if (nickname.isBlank()) {
			throw new BusinessException(ErrorCode.MISSING_REQUIRED_FIELD);
		}
		if (nickname.length() > 50) {
			throw new BusinessException(ErrorCode.INVALID_REQUEST_DATA);
		}
	}

	// 핸드폰번호 검증 메서드
	private void validatePhoneNumber(String phoneNumber) {
		if (phoneNumber.isBlank()) {
			throw new BusinessException(ErrorCode.MISSING_REQUIRED_FIELD);
		}

		String phoneRegex = "^010-\\d{4}-\\d{4}$";
		if (!phoneNumber.matches(phoneRegex)) {
			throw new BusinessException(ErrorCode.INVALID_REQUEST_DATA);
		}
	}
}

File: Readio/src/main/java/ninegle/Readio/mypage/service/MyPageUserService.kt
package ninegle.Readio.mypage.service

import ninegle.Readio.global.exception.BusinessException
import ninegle.Readio.global.exception.domain.ErrorCode
import ninegle.Readio.mypage.dto.request.UserUpdateRequestDto
import ninegle.Readio.mypage.dto.response.UserInfoDto
import ninegle.Readio.mypage.mapper.MyPageUserMapper
import ninegle.Readio.user.repository.UserRepository
import ninegle.Readio.user.service.UserContextService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

/**
 * 마이페이지 사용자 정보 관리 서비스
 */
@Service
class MyPageUserService(
	private val userRepository: UserRepository,
	private val userContextService: UserContextService
) {

	companion object {
		private const val MAX_NICKNAME_LENGTH = 50
		private val PHONE_NUMBER_REGEX = Regex("^010-\\d{4}-\\d{4}$")
	}

	//사용자 정보 조회
	@Transactional(readOnly = true)
	fun getUserInfo(): UserInfoDto {
		val userId = userContextService.getCurrentUserId()

		val user = userRepository.findById(userId).orElseThrow {
			BusinessException(ErrorCode.USER_NOT_FOUND)
		}

		return MyPageUserMapper.toUserInfoDto(user)
	}

	//사용자 정보 업데이트
	@Transactional
	fun updateUserInfo(dto: UserUpdateRequestDto): UserInfoDto {
		val userId = userContextService.getCurrentUserId()

		// 닉네임, 전화번호 둘 다 비어있는 경우 예외 처리
		if (dto.nickname.isNullOrBlank() && dto.phoneNumber.isNullOrBlank()) {
			throw BusinessException(ErrorCode.MISSING_REQUIRED_FIELD)
		}

		val user = userRepository.findById(userId).orElseThrow {
			BusinessException(ErrorCode.USER_NOT_FOUND)
		}

		// 닉네임 업데이트 로직
		dto.nickname?.let { nickname ->
			validateNickname(nickname)
			user.updateNickname(nickname)
		}

		// 전화번호 업데이트 로직
		dto.phoneNumber?.let { phoneNumber ->
			validatePhoneNumber(phoneNumber)
			user.updatePhoneNumber(phoneNumber)
		}

		return MyPageUserMapper.toUserInfoDto(user)
	}

	//닉네임 유효성 검증 메서드
	private fun validateNickname(nickname: String) {
		if (nickname.isBlank()) {
			throw BusinessException(ErrorCode.MISSING_REQUIRED_FIELD)
		}
		if (nickname.length > MAX_NICKNAME_LENGTH) {
			throw BusinessException(ErrorCode.INVALID_REQUEST_DATA)
		}
	}

	//핸드폰번호 유효성 검증 메서드
	private fun validatePhoneNumber(phoneNumber: String) {
		if (phoneNumber.isBlank()) {
			throw BusinessException(ErrorCode.MISSING_REQUIRED_FIELD)
		}

		if (!phoneNumber.matches(PHONE_NUMBER_REGEX)) {
			throw BusinessException(ErrorCode.INVALID_REQUEST_DATA)
		}
	}
}